### PR TITLE
dcn on Criteo 

### PR DESCRIPTION
**Pull request by jiangyzy**

### CODE REVIEW

File: RecommenderSystems/dcn/dcn_train_eval.py [hunk 1]
```diff
@@ -0,0 +1,857 @@
...
+import sys
+import glob
+import time
+import math
+import numpy as np
+import psutil
+import warnings
+import oneflow as flow
+import oneflow.nn as nn
+from sklearn.metrics import roc_auc_score, log_loss
```
Comment (ShawnXuan) [inline RecommenderSystems/dcn/dcn_train_eval.py:13]: replace roc_auc_score with flow.roc_auc_score
is it possible to replace log_loss with flow.nn.BCELoss?

File: RecommenderSystems/dcn/dcn_train_eval.py [hunk 2]
```diff
@@ -0,0 +1,857 @@
...
+    batches_per_epoch = math.ceil(args.num_train_samples / args.train_batch_size)
+    batches_per_epoch = 3667
+
+    test_logloss_best = np.inf
+    test_auc_best = 0
+    best_metric = -np.inf
+    stopping_steps = 0
+
+    dcn_module.train()
+    step, last_step, last_time = -1, 0, time.time()
```
Comment (ShawnXuan) [inline RecommenderSystems/dcn/dcn_train_eval.py:625]: move to L632

File: RecommenderSystems/dcn/dcn_train_eval.py [hunk 3]
```diff
@@ -0,0 +1,857 @@
...
+        if i == 1:
+            break
+
+    print("best test auc : ", test_auc_best, "logloss: ", test_logloss_best)
+
+    plot_train_curve(args, train_losses)
+    np.save("of_train_losses.npy", train_losses)
+
+
+def batch_to_global(np_label, np_features):
```
Comment (ShawnXuan) [inline RecommenderSystems/dcn/dcn_train_eval.py:701]: 确认一下数据类型，这里可以节省cast

**Review by ShawnXuan — COMMENTED** [review]
tools下面的文件未来只需保留给oneflow制作数据集的脚本即可

File: RecommenderSystems/dcn/train.sh [hunk 4]
```diff
@@ -0,0 +1,40 @@
+#!/bin/bash
+rm -r /home/yuanziyang/yzywork/dcn-test-dir/persistent1/*
```
Comment (ShawnXuan) [inline RecommenderSystems/dcn/train.sh:2]: use a fake path

File: RecommenderSystems/dcn/dcn_train_eval.py (hunk 1 above)
Comment (jiangyzy) [reply RecommenderSystems/dcn/dcn_train_eval.py:13]: 已更改

File: RecommenderSystems/dcn/dcn_train_eval.py [hunk 5]
```diff
@@ -0,0 +1,832 @@
...
+        cache_memory_budget_mb=args.cache_memory_budget_mb,
+        dnn_hidden_units=args.dnn_hidden_units,
+        crossing_layers=args.crossing_layers,
+        net_dropout=args.net_dropout,
+        dnn_activations=args.dnn_activations,
+        batch_norm=args.batch_norm,
+        size_factor=args.size_factor,
+    )
+    for key in model.state_dict():
+        print(key)
```
Comment (ShawnXuan) [inline RecommenderSystems/dcn/dcn_train_eval.py:448]: unnecessary

File: RecommenderSystems/dcn/dcn_train_eval.py [hunk 6]
```diff
@@ -0,0 +1,832 @@
...
+
+
+def plot_train_curve(args, train_losses):
+    plt.plot(range(1, len(train_losses) + 1), train_losses)
+    plt.ylabel("Training Logloss")
+    plt.xlabel(f"batch number (per {args.loss_print_interval} batches)")
+    plt.savefig("of_training_curve.png")
+
+
+def load_fuxi_test(args):
```
Comment (ShawnXuan) [inline RecommenderSystems/dcn/dcn_train_eval.py:793]: unnecessary

File: RecommenderSystems/dcn/dcn_train_eval.py [hunk 7]
```diff
@@ -0,0 +1,832 @@
...
+    return table_size_array
+
+
+if __name__ == "__main__":
+    os.system(sys.executable + " -m oneflow --doctor")
+    flow.boxing.nccl.enable_all_to_all(True)
+    args = get_args()
+    args.table_size_array = get_table_size_array(args.feature_map_json)
+    train(args)
+    # load_fuxi_test(args)
```
Comment (ShawnXuan) [inline RecommenderSystems/dcn/dcn_train_eval.py:832]: unnecessary

File: RecommenderSystems/dcn/dcn_train_eval.py [hunk 8]
```diff
@@ -0,0 +1,832 @@
...
+                        opt.state_dict()["param_groups"][0]["_options"]["lr"],
+                    )
+                    epoch += 1
+                    auc, logloss = eval_valid(args, eval_graph, step, epoch)
+                    test_auc, test_logloss = eval_test(args, eval_graph, step, epoch)
+                    if test_auc > test_auc_best:
+                        test_auc_best = test_auc
+                        test_logloss_best = test_logloss
+
+                    # if args.save_model_after_each_eval:
```
Comment (ShawnXuan) [inline RecommenderSystems/dcn/dcn_train_eval.py:640]: save model after each eval

File: RecommenderSystems/dcn/dcn_train_eval.py [hunk 9]
```diff
@@ -0,0 +1,832 @@
...
+    parser.add_argument("--model_load_dir", type=str, default=None)
+    parser.add_argument("--model_save_dir", type=str, default=None)
+    parser.add_argument(
+        "--save_initial_model", action="store_true", help="save initial model parameters or not."
+    )
+    parser.add_argument(
+        "--save_model_after_each_eval", action="store_true", help="save model after each eval."
+    )
+
+    parser.add_argument("--disable_fusedmlp", default=False, help="disable fused MLP or not")
```
Comment (ShawnXuan) [inline RecommenderSystems/dcn/dcn_train_eval.py:63]: usless

File: RecommenderSystems/dcn/dcn_train_eval.py [hunk 10]
```diff
@@ -0,0 +1,832 @@
...
+        table_size_array.append(feature_specs[col]["vocab_size"])
+
+    return table_size_array
+
+
+if __name__ == "__main__":
+    os.system(sys.executable + " -m oneflow --doctor")
+    flow.boxing.nccl.enable_all_to_all(True)
+    args = get_args()
+    args.table_size_array = get_table_size_array(args.feature_map_json)
```
Comment (ShawnXuan) [inline RecommenderSystems/dcn/dcn_train_eval.py:830]: table_size_array is overwrited

Comment (ShawnXuan) [reply RecommenderSystems/dcn/dcn_train_eval.py:830]: It's a good idea to use "feature_map.json", but currently I suggest to use table_size_array only.  We consider using it in the future.

File: RecommenderSystems/dcn/dcn_train_eval.py [hunk 11]
```diff
@@ -0,0 +1,695 @@
...
+
+    def get_batches(self, reader, batch_size=None):
+        if batch_size is None:
+            batch_size = self.batch_size
+
+        tail = None
+
+        for rg in reader:
+            rgdict = rg._asdict()
+            rglist = np.array([rgdict[field] for field in self.fields])
```
Comment (ShawnXuan) [inline RecommenderSystems/dcn/dcn_train_eval.py:155]: np.array will force to convert dtype to float64.
more information: 
https://github.com/Oneflow-Inc/models/pull/339/files#r869078975

File: RecommenderSystems/dcn/dcn_train_eval.py [hunk 12]
```diff
@@ -0,0 +1,695 @@
...
+            
+
+    print("best test auc : ", test_auc_best, "logloss: ", test_logloss_best)
+
+
+
+
+def batch_to_global(np_label, np_features):
+    def _np_to_global(np, dtype=flow.float):
+        t = flow.tensor(np, dtype=dtype)
```
Comment (ShawnXuan) [inline RecommenderSystems/dcn/dcn_train_eval.py:632]: use `from_numpy`

File: RecommenderSystems/dcn/dcn_train_eval.py [hunk 13]
```diff
@@ -0,0 +1,695 @@
...
+        return t.to_global(placement=flow.env.all_device_placement("cpu"), sbp=flow.sbp.split(0))
+
+    labels = _np_to_global(np_label.reshape(-1, 1))
+    features = _np_to_global(np_features, dtype=flow.int64)
+    return labels, features
+
+
+def eval(args, eval_graph, tag='val',  cur_step=0, epoch=0):
```
Comment (ShawnXuan) [inline RecommenderSystems/dcn/dcn_train_eval.py:640]: ref deepfm eval
- log_loss
- labels, preds
- ...

File: RecommenderSystems/dcn/train.sh [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+      --eval_batch_size 10000 \
+      --eval_batches 1000 \
+      --dnn_hidden_units "1000, 1000, 1000, 1000, 1000" \
+      --crossing_layers 4\
+      --net_dropout 0.2 \
+      --learning_rate 0.001 \
+      --embedding_vec_size 16 \
+      --num_train_samples 36672493 \
+      --size_factor 3 > run.log 2>&1 & 
+      tail -f run.log
```
Comment (ShawnXuan) [inline RecommenderSystems/dcn/train.sh:37]: use tee or nothing

File: RecommenderSystems/dcn/train.sh [hunk 15]
```diff
@@ -0,0 +1,37 @@
+export CUDA_VISIBLE_DEVICES=0
```
Comment (ShawnXuan) [inline RecommenderSystems/dcn/train.sh:1]: usless

File: RecommenderSystems/dcn/dcn_train_eval.py [hunk 16]
```diff
@@ -0,0 +1,695 @@
...
+    stopping_steps = 0
+    stop_training = False
+
+    with make_criteo_dataloader(f"{args.data_dir}/train", args.train_batch_size, shard_seed=args.shard_seed) as loader:
+        dcn_module.train()
+        last_step, last_time = 0, time.time()
+        for step in range(1, args.train_batches + 1):
+            labels, features = batch_to_global(*next(loader))
+            loss = train_graph(labels, features)
+            loss = loss.numpy()
```
Comment (ShawnXuan) [inline RecommenderSystems/dcn/dcn_train_eval.py:579]: mv loss.numpy() after line 581

File: RecommenderSystems/dcn/dcn_train_eval.py [hunk 17]
```diff
@@ -0,0 +1,686 @@
+import argparse
+from multiprocessing.spawn import get_executable
```
Comment (ShawnXuan) [inline RecommenderSystems/dcn/dcn_train_eval.py:2]: useless

File: RecommenderSystems/dcn/dcn_train_eval.py [hunk 18]
```diff
@@ -0,0 +1,686 @@
...
+import sys
+import glob
+import time
+import math
+import numpy as np
+import psutil
+import oneflow as flow
+import oneflow.nn as nn
+from petastorm.reader import make_batch_reader
+from multiprocessing.spawn import get_executable
```
Comment (ShawnXuan) [inline RecommenderSystems/dcn/dcn_train_eval.py:13]: useless

**Review by ShawnXuan — COMMENTED** [review]
remove RecommenderSystems/dcn/tools/make_criteo_parquet.py and RecommenderSystems/dcn/tools/dataset_config.yaml

File: RecommenderSystems/dcn/dcn_train_eval.py [hunk 19]
```diff
@@ -0,0 +1,700 @@
...
+        if not isinstance(dropout_rates, list):
+            dropout_rates = [dropout_rates] * len(hidden_units)
+        hidden_activations = [nn.ReLU()] * len(hidden_units)
+        hidden_units = [input_dim] + hidden_units
+        for idx in range(len(hidden_units) - 1):
+            dense_layers.append(nn.Linear(hidden_units[idx], hidden_units[idx + 1], bias=use_bias))
+            if batch_norm:
+                dense_layers.append(nn.BatchNorm1d(hidden_units[idx + 1]))
+            if hidden_activations[idx]:
+                dense_layers.append(hidden_activations[idx])
```
Comment (ShawnXuan) [inline RecommenderSystems/dcn/dcn_train_eval.py:309]: dense_layers.append(nn.ReLU())

File: RecommenderSystems/dcn/dcn_train_eval.py [hunk 20]
```diff
@@ -0,0 +1,700 @@
...
+        flat_feature_emb = feature_emb.flatten(start_dim=1)
+        cross_out = self.crossnet(flat_feature_emb)
+        if self.dnn is not None:
+            dnn_out = self.dnn(flat_feature_emb)
+            final_out = flow.cat([cross_out, dnn_out], dim=-1)
+        else:
+            final_out = cross_out
+        y_pred = self.fc(final_out)
+        y_pred = self.output_activation(y_pred)
+        return y_pred
```
Comment (ShawnXuan) [inline RecommenderSystems/dcn/dcn_train_eval.py:380]: prefer return y_pred.sigmoid()

File: RecommenderSystems/dcn/dcn_train_eval.py [hunk 21]
```diff
@@ -0,0 +1,700 @@
...
+class DCNValGraph(flow.nn.Graph):
+    def __init__(self, dcn_module, amp=False):
+        super(DCNValGraph, self).__init__()
+        self.module = dcn_module
+        if amp:
+            self.config.enable_amp(True)
+
+    def build(self, features):
+        predicts = self.module(features.to("cuda"))
+        return predicts.to("cpu")
```
Comment (ShawnXuan) [inline RecommenderSystems/dcn/dcn_train_eval.py:417]: return predicts

File: RecommenderSystems/dcn/dcn_train_eval.py [hunk 22]
```diff
@@ -0,0 +1,700 @@
...
+            if step % batches_per_epoch == 0:
+                epoch += 1
+                auc, logloss = eval(
+                    args,
+                    eval_graph,
+                    tag="val",
+                    cur_step=step,
+                    epoch=epoch,
+                    cached_eval_batches=cached_valid_batches,
+                )
```
Comment (ShawnXuan) [inline RecommenderSystems/dcn/dcn_train_eval.py:574]: test only happens at the end of training

File: RecommenderSystems/dcn/README.md [hunk 23]
```diff
@@ -0,0 +1,178 @@
+# Deep&Cross
+ [Deep & Cross Network](https://dl.acm.org/doi/10.1145/3124749.3124754) (DCN) can not only keep the advantages of DNN model, but also learn specific bounded feature crossover more effectively. In particular, DCN can explicitly learn cross features for each layer without the need for manual feature engineering, and the increased algorithm complexity is almost negligible compared with DNN model.
+
+
+## Directory description
+```
+.
+|-- tools
+  |-- dataset_config.yaml   # dataset config file
```
Comment (ShawnXuan) [inline RecommenderSystems/dcn/README.md:9]: dcn_parquet.scala

File: RecommenderSystems/dcn/README.md [hunk 24]
```diff
@@ -0,0 +1,178 @@
...
+
+     -   Modify the SPARK_LOCAL_DIRS as needed
+
+         ```shell
+         export SPARK_LOCAL_DIRS=/path/to/your/spark/
+         ```
+
+     -   Run `bash launch_spark.sh`
+
+3. load [ddn_parquet.scala](https://github.com/Oneflow-Inc/models/blob/criteo_dcn/RecommenderSystems/dcn/tools/dcn_parquet.scala) to your spark shell by `:load ddn_parquet.scala`.
```
Comment (ShawnXuan) [inline RecommenderSystems/dcn/README.md:107]: dcn_parquet.scala

File: RecommenderSystems/dcn/README.md [hunk 25]
```diff
@@ -0,0 +1,167 @@
+# Deep&Cross
+ [Deep & Cross Network](https://dl.acm.org/doi/10.1145/3124749.3124754) (DCN) can not only keep the advantages of DNN model, but also learn specific bounded feature crossover more effectively. In particular, DCN can explicitly learn cross features for each layer without the need for manual feature engineering, and the increased algorithm complexity is almost negligible compared with DNN model.
+ ![DCN](https://user-images.githubusercontent.com/80230303/159417248-1975736f-3de8-4972-84e3-2f0f346cbc1a.png)
+
+
+Oneflow API is compatible to Pytorch, so only minor modification in codes then we can apply the Pytorch implemented modules to Oneflow. Therefore, We refer to some DCN implementation in [FuxiCTR](https://github.com/xue-pai/FuxiCTR/tree/v1.0.2) such as CrossNet. We follow the FuxiCTR [CrossNet implementation](https://github.com/xue-pai/FuxiCTR/blob/0f4ccf36aac02d120cb0c87f417c40e0e4330744/fuxictr/pytorch/models/DCN.py#L77), reuse the `CrossInteractionLayer` as a basic Crossnet layer, and make it multi layers in `CrossNet`. The only difference of CrossNet implementation between FuxiCTR and Oneflow is `nn` module, one is `torch.nn` and another one is `oneflow.nn` .
```
Comment (ShawnXuan) [inline RecommenderSystems/dcn/README.md:6]: Therefore, we adopted some implementation from [FuxiCTR](https://github.com/xue-pai/FuxiCTR/tree/v1.0.2), for example the `CrossInteractionLayer` was reused as a basic `CrossNet` hidden layer, and we make a recurrent loop of these hidden layers in `CrossNet` module to get high-degree interaction across features.

~The only difference of CrossNet implementation between FuxiCTR and Oneflow is `nn` module, one is `torch.nn` and another one is `oneflow.nn` .~

**Review by ShawnXuan — APPROVED** [review]
